libgalois/include/katana/GraphTopology.h

```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    KATANA_LOG_DEBUG_ASSERT(static_cast<size_t>(d) < num_nodes());
+    KATANA_LOG_DEBUG_ASSERT(d > 0);
+
+    return static_cast<Node>(d);
+  }
+
+  Node edge_dest(Edge edge_id) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(edge_id < dests_.size());
+    return dests_[edge_id];
+  }
+
+  nodes_range nodes(Node begin, Node end) const noexcept {
+    return MakeStandardRange<node_iterator>(begin, end);
+  }
+
+  nodes_range all_nodes() const noexcept {
+    return nodes(Node{0}, static_cast<Node>(num_nodes()));
+  }
+
+  edges_range all_edges() const noexcept {
+    return MakeStandardRange<edge_iterator>(Edge{0}, Edge{num_edges()});
+  }
+  // Standard container concepts
+
+  node_iterator begin() const noexcept { return node_iterator(0); }
+
+  node_iterator end() const noexcept { return node_iterator(num_nodes()); }
+
+  size_t size() const noexcept { return num_nodes(); }
+
+  bool empty() const noexcept { return num_nodes() == 0; }
+
+  ///@param node node to get degree for
+  ///@returns Degree of node N
+  size_t degree(Node node) const noexcept { return edges(node).size(); }
+
+  PropertyIndex edge_property_index(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < num_edges());
+    return new_to_old_edges_mapping_[eid];
+  }
+
+  Edge original_edge_id(const Edge& eid) const noexcept {
+    return edge_property_index(eid);
+  }
+
+  Edge projected_edge_id(const Edge& eid) const noexcept {
+    return old_to_new_edges_mapping_[eid];
+  }
+
+  PropertyIndex node_property_index(const Node& nid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(nid < num_nodes());
+    return new_to_old_nodes_mapping_[nid];
+  }
+
+  Node original_node_id(const Node& nid) const noexcept {
+    return node_property_index(nid);
+  }
+
+  Node projected_node_id(const Node& nid) const noexcept {
+    return old_to_new_nodes_mapping_[nid];
+  }
+
+  static std::unique_ptr<ProjectedShuffleTopology> Make(
+      const PropertyGraph* pg, const std::vector<std::string>& node_types,
+      const std::vector<std::string>& edge_types);
+
+private:
+  ProjectedShuffleTopology(
+      NUMAArray<Edge>&& adj_indices, NUMAArray<Node>&& dests,
+      NUMAArray<Node>&& old_to_new_nodes_mapping,
+      NUMAArray<Node>&& new_to_old_nodes_mapping,
+      NUMAArray<Edge>&& old_to_new_edges_mapping,
+      NUMAArray<Edge>&& new_to_old_edges_mapping) noexcept
+      : adj_indices_(std::move(adj_indices)),
+        dests_(std::move(dests)),
+        old_to_new_nodes_mapping_(std::move(old_to_new_nodes_mapping)),
+        new_to_old_nodes_mapping_(std::move(new_to_old_nodes_mapping)),
+        old_to_new_edges_mapping_(std::move(old_to_new_edges_mapping)),
+        new_to_old_edges_mapping_(std::move(new_to_old_edges_mapping)) {}
+
+  NUMAArray<Edge> adj_indices_;
+  NUMAArray<Node> dests_;
+  NUMAArray<Node> old_to_new_nodes_mapping_;
+  NUMAArray<Node> new_to_old_nodes_mapping_;
+  NUMAArray<Edge> old_to_new_edges_mapping_;
```

Comment:
Isn't this the size of the edges (and vertices above) in the original graph?

Do we need this? This is a trade-off. We can do binary search in `new_to_old_` instead. For this PR, this is fine. Please add a TODO to revisit this (also describe the trade-off).